dimod/include/dimod/constrained_quadratic_model.h

```diff
@@ -0,0 +1,493 @@
+    objective.reindex_variables(v);
+    varinfo_.erase(varinfo_.begin() + v);
+}
+
+template <class bias_type, class index_type>
+void ConstrainedQuadraticModel<bias_type, index_type>::set_lower_bound(index_type v, bias_type lb) {
+    varinfo_[v].lb = lb;
+}
+
+template <class bias_type, class index_type>
+template <class B, class I>
+void ConstrainedQuadraticModel<bias_type, index_type>::set_objective(
+        const abc::QuadraticModelBase<B, I>& objective) {
+    // todo: move version
+    // todo: support sparse mapping
+
+    // add missing variables
+    for (size_type v = num_variables(); v < objective.num_variables(); ++v) {
+        add_variable(objective.vartype(v), objective.lower_bound(v), objective.upper_bound(v));
+    }
+
+    this->objective.clear();
+
+    // linear biases
+    for (size_type v = 0; v < objective.num_variables(); ++v) {
+        // this checks the overlapping variables. todo: do this in a NDEBUG ifdef?
+        assert(vartype(v) == objective.vartype(v));
+        assert(lower_bound(v) == objective.lower_bound(v));
+        assert(upper_bound(v) == objective.upper_bound(v));
+
+        this->objective.add_linear(v, objective.linear(v));
+    }
+
+    // quadratic biases
+    for (auto it = objective.cbegin_quadratic(); it != objective.cend_quadratic(); ++it) {
+        this->objective.add_quadratic(it->u, it->v, it->bias);
+    }
+
+    // offset
+    this->objective.add_offset(objective.offset());
+}
+
+// variables must all be present in this case!
+template <class bias_type, class index_type>
+template <class B, class I, class T>
+void ConstrainedQuadraticModel<bias_type, index_type>::set_objective(
+        const abc::QuadraticModelBase<B, I>& objective, const std::vector<T>& mapping) {
+    // there are a tonne of potential optimizations here, but let's keep it simple for now
+    assert(mapping.size() == objective.num_variables());
+
+    this->objective.clear();
+
+    for (size_type i = 0; i < objective.num_variables(); ++i) {
+        assert(vartype(mapping[i]) == objective.vartype(i));
+        assert(lower_bound(mapping[i]) == objective.lower_bound(i));
+        assert(upper_bound(mapping[i]) == objective.upper_bound(i));
+
+        this->objective.add_linear(mapping[i], objective.linear(i));
+    }
+
+    // quadratic biases
+    for (auto it = objective.cbegin_quadratic(); it != objective.cend_quadratic(); ++it) {
+        this->objective.add_quadratic(mapping[it->u], mapping[it->v], it->bias);
+    }
+
+    // offset
+    this->objective.add_offset(objective.offset());
+}
+
+template <class bias_type, class index_type>
+void ConstrainedQuadraticModel<bias_type, index_type>::set_upper_bound(index_type v, bias_type ub) {
+    varinfo_[v].ub = ub;
+}
+
+template <class bias_type, class index_type>
+void ConstrainedQuadraticModel<bias_type, index_type>::set_vartype(index_type v, Vartype vartype) {
+    varinfo_[v].vartype = vartype;
```

Comment:
need asserts ? , like if the old ub and lb are still valid or not.

Reply:
In this case, no, because the user might want to temporarily make an invalid model. Applies to the others.